### ISSUE TITLE
Displacement and Normed Displacement

### ISSUE THREAD

**Issue by acse-tz1023** [dl4to/dl4to#5]
Dear DL4TO team,

This is Terry again, I have 2 more questions regarding the use of DL4TO:
1. I set up the problem class, and then I assign the FDM() as the pde_solver to this problem.  After that, I set up the criterion to be compliance and a VolumeConstraint(max_volumefraction = .21, thresold_fct = 'relu'). SIMP using this criterion with binarizer_steepening_factor = 1.02, and 500 iterations, lr = 1e-3.
Then I plot the solution, the strange thing is if I modify the max_volume_fraction to be .17, then the simp will almost not work. With .21, I can see that SIMP returns me a 10-cell potential optimized topology, but with .17, it will be stuck at the first cell. The SIMP will not try adding cells to make the structure better.
Do you have any theory why this is happening? If this behavior is correct, could you give me some advice on how should I tune the parameter in SIMP?

2. the force direction plotting seems not working on my end, is that a known issue?

Thank you in advance for your time and patience. I look forward to your response.

Best,
Terry

**Comment by dl4to** [comment]
Dear Terry,

1. There are several things that might be happening here. Since I don't know the specifications of the TO problem you're trying to solve, I can only make guesses.

First, it's possible that your TO problem constrains the densities in some way. For example, you might restrict the density of some voxels to be 1. This happens automatically for voxels with a force applied to them. To check this, please look at the "Design space information" plot in problem.plot(). This can be problematic because, by default, the volume fraction in the VolumeConstraint criterion is calculated over the entire domain. Therefore, your volume fraction of 17% (in your example) might already be exceeded just from the voxels forced to be 1 by your problem formulation. Note that the VolumeConstraint criterion has an argument called compute_only_on_design_space (which is False by default). If you set this to True, the volume fraction is computed only for voxels that are free to optimize in your problem formulation.

I just realized that enforcing a density of 1 for voxels with an applied force can be problematic in cases of volumetric forces applied everywhere (such as gravity). I will update the code to add an argument to the Problem class (called "restrict_density_for_voxels_with_applied_forces" or something similar) to toggle this behavior.

Second, it's possible that the density values in your optimization are below the threshold of 0.5. Please plot the solution with the argument binary=False to check if this is the case.

2. Regarding the force direction plotting, it is working on my end, so I'm not sure why it doesn't work for you. I assume this might be due to some version issues with Plotly. I have version 5.9.0 installed.

Best regards,
David

**Comment by dl4to** [comment]
I would recommend setting Ω_design to be 1 where Ω_dirichlet is 1, just to be sure. You can do that with: 
```
Ω_design = torch.zeros(1, *shape)
Ω_design[0, :, :, 1] = -1
Ω_design[0, 0, 0, 1] = 1
Ω_design[0, -1, :-1, 1] = 1
```
Furthermore, I would probably use a padding depth of 1 for these low-resolution examples. Then at least in my experiments you get a somewhat viable solution for a volume fraction of 0.20.

I should add that the results are a bit unreliable for such extremely low-resolution problems. You get much better examples if you increase the resolution. For instance, try shape = [10, 10, 3].

In any case however, you don't get good results for volume fractions of 0.17. This seems to be too restrictive for the optimization.

By the way, you can easily check the volume of your structure with solution.get_θ(binary=True).mean().

This is the code that I used for your example:

```
import torch
from dl4to.problem import Problem
from dl4to.pde import FDM
from dl4to.criteria import Compliance, VolumeConstraint
from dl4to.topo_solvers import SIMP

E = 2e11
ν = .3
σ_ys = 4.5e10

shape = [10, 10, 3]
h = [0.250, 0.250, 0.250]

# define external forces
F = torch.zeros(3, *shape)
F[1, -1, -1, 1] = -1e5

Ω_dirichlet = torch.zeros(3, *shape)
Ω_dirichlet[1, 0, 0, 1] = 1
Ω_dirichlet[0, -1, :-1, 1] = 1

Ω_design = torch.zeros(1, *shape)
Ω_design[0, :, :, 1] = -1
Ω_design[0, 0, 0, 1] = 1
Ω_design[0, -1, :-1, 1] = 1

problem = Problem(E, ν, σ_ys, h, Ω_dirichlet, Ω_design, F)
problem.pde_solver = FDM(padding_depth=0) # or padding_depth=1

criterion = Compliance() + VolumeConstraint(max_volume_fraction=0.19, threshold_fct="relu")

simp = SIMP(
    criterion=criterion,
    binarizer_steepening_factor=1.02,
    n_iterations=500,
    lr=1e-3)

solution = simp(problem)

print("Volume", solution.get_θ(binary=True).mean().item())

camera_position = (0, 1.7, 1.1)
solution.plot(camera_position=camera_position,
              binary=False,
              solve_pde=True)
```

**Comment by acse-tz1023** [comment]
Thanks! I think you are right about how restrictive my settings are, so I changed it to a finer resolution, and it works well now.
I got one more question about the gt_solution in the dataset provided in DL4TO, I found that for some example, the compliance of the Unet3D topo_solver is better than the gt_solution and the SIMP solution, and this is based on the UNet criterion is weighted BCE instead of compliance. Furthermore, the design of the UNet resulting design seems to be even closer to the gt_solution than the SIMP result. 
What I am confused about is why the Unet result is so good, I mean theoretically machine learning should not achieve a better result than the gradient based method such as SIMP.
In case you would wonder what code I executed,
`from dl4to.criteria import Compliance, VolumeConstraint
from dl4to.solution import Solution
from dl4to.datasets import SELTODataset
from dl4to.utils import get_dataloader
from dl4to.preprocessing import TrivialPreprocessing
from dl4to.criteria import WeightedBCE
from dl4to.topo_solvers import TrainableTopoSolver
from torch.optim import Adam
from dl4to.models import UNet3D
from torchinfo import summary
import torch

criterion = Compliance() + VolumeConstraint(max_volume_fraction=.12, threshold_fct='relu')

from dl4to.pde import FDM

dataset_train = SELTODataset("/home/tz1023/virtual_local/IRP/DL4TOlearn/data/", name="disc_simple", train=True)
dataset_val = SELTODataset("/home/tz1023/virtual_local/IRP/DL4TOlearn/data/", name="disc_simple", train=False)
problem, gt_solution = dataset_val[2]

problem.pde_solver = FDM(padding_depth=0)

camera_position = [0,0.02,0.12]
problem.plot(camera_position=camera_position,
             show_axislabels=True,
             show_ticklabels=True,
             display=False)

dataloader_train = get_dataloader(dataset_train, batch_size=128)
dataloader_val = get_dataloader(dataset_val, batch_size=1)

preprocessing = TrivialPreprocessing(normalize=True, dataset=dataset_train)
preprocessed_problem = preprocessing(problem)

from dl4to.topo_solvers import SIMP

simp = SIMP(
    criterion=criterion,
    binarizer_steepening_factor=1.02,
    n_iterations=200,
    lr=3e-1,
)

solutionSIMP = simp(problem)  # In order to apply a topo_solver to a problem, we can simply call it via solution = topo_solver(problem), which returns a solution object
camera_position = (0, 0.06, 0.12)
solutionSIMP.plot(camera_position=camera_position,
              binary=True,
              show_axislabels=True,
              show_ticklabels=True,
              solve_pde=True,
              display=True)

from dl4to.criteria import VolumeFraction

volume_fraction = VolumeFraction()

criterion = WeightedBCE()
criterion.set_optimal_weight(dataset_train)

train_model = False   # Control if the model should be re-trained or not
model = UNet3D(in_channels=preprocessing.shape, pooling_kernel_size=[2,2,1]).to('cpu')
optimizer = Adam(model.parameters(), lr=1e-3)

if train_model:
    topo_solver = TrainableTopoSolver(criterion=criterion,
                                    model=model,
                                    optimizer=optimizer,
                                    preprocessing=preprocessing)

    summary(model, input_size=(1, preprocessing.shape, 39, 39, 4))  # visualize the model

    topo_solver.train(
            root='training/unet',
            epochs=40,
            dataloader_train=dataloader_train,
            dataloader_val=dataloader_val,
            validation_interval=10
    )
    torch.save({
    'model_state_dict': model.state_dict(),
    'optimizer_state_dict': optimizer.state_dict(),
    'epoch': 40,
    }, 'topo_solver_checkpoint.pth')
else:
    checkpoint = torch.load('topo_solver_checkpoint.pth')

    # Load the model and optimizer states
    model.load_state_dict(checkpoint['model_state_dict'])
    optimizer.load_state_dict(checkpoint['optimizer_state_dict'])

    # Reinitialize topo_solver with the loaded model and optimizer
    topo_solver = TrainableTopoSolver(
        criterion=criterion,
        model=model,
        optimizer=optimizer,
        preprocessing=preprocessing
    )


solutionML = topo_solver(problem)
camera_position = [0,0.02,0.12]
solutionML.plot(threshold=.5, 
              camera_position=camera_position,
              display=True)

gt_solution.plot(threshold=.5, 
              camera_position=camera_position,
              display=True)

print('solutionSIMP: ',volume_fraction(solutionSIMP))
print('solutionML: ',volume_fraction(solutionML))
print('solutionGT: ',volume_fraction(gt_solution))`

**Comment by dl4to** [comment]
1. Your UNet model seems to output designs that have a bit more volume than the ground truth solutions (compare solutionML.θ.mean() to the other solutions). Therefore the compliance is better, but the overall SIMP objective function (which is compliance plus volume constraint) is not necessarily better. Why this happens is not easy to say, maybe coincidentally the solutions in the training dataset have more volume on average and the model picked this up. But it is often hard to interpret why deep learning models behave a certain way.
2. The ground truth solutions have been generated in cooperation with the Ariane Group (https://www.ariane.group/en/) and Synera (https://www.synera.io/) using the Altair OptiStruct (https://www.altair.com/optistruct/) implementation of SIMP within the Synera software. This is because when I generated the SELTO datasets my own SIMP implementation was not working yet. Today I would use the SIMP implementation in the DL4TO library to generate the ground truths, but they are very similar to the OptiStruct solutions anyway. But that is the reason why SIMP solutions that have been generated using DL4TO sometimes look a little bit different to the ground truth solutions (which, as I said earlier, have been generated using OptiStruct).

**Comment by acse-tz1023** [comment]
For the first answer, I have a follow-up question: if the volumes are the same or very close for the Deep learning model and SIMP model, should SIMP always have better compliance? I can check the volume of the solutionML and then manually set that to be the volume constraint of SIMP.

Is it correct to check the compliance of gt_solution by calling compliance(gt_solution) like I did in the earlier question? I found the print message of the last 3 lines of the code are 0.0017, 0.0016 and 0.0576 if solving for problem in dataset_val[2].

**Comment by dl4to** [comment]
> if the volumes are the same or very close for the Deep learning model and SIMP model, should SIMP always have better compliance?

I thank that question cannot be answered universally since it always depends on the parameter settings and training dataset. But generally I would say that yes, SIMP is expected to be better because it explicitly runs an optimization run for that specific problem whereas the neural network has not seen that problem before and only learns from similar problems from the training dataset. So SIMP has more chances to be closer to the optimum for that problem, but at the cost of a longer inference time

> Is it correct to check the compliance of gt_solution by calling compliance(gt_solution) like I did in the earlier question?

Yes that is correct.

> Additionally, do you mind sharing the criterion that you used to for SIMP to generate the current ground truth solution?

I unfortunately don't have access to the exact criterion that was used to generate the ground truths, because it was not me who generated the solutions (that was done by engineers at Ariane Group).
